### PR TITLE
[SYCL] Cleanup/refactor device_code_tests

### PR DESCRIPTION
**Pull request by ianayl**
This PR:
- Changes more lit tests in `sycl/test/check_device_code` to use `SYCL_EXTERNAL` functions in order to remove redundant host-side boilerplate code, and
- Moves tests not verifying device code outside of the `check_device_code` folder.
- Moves tests verifying device code that are outside of the  `check_device_code` folder into the `check_device_code` folder

### CODE REVIEW

File: sycl/test/check_device_code/ap_fixed.cpp
Comment (aelovikov-intel) [inline sycl/test/check_device_code/ap_fixed.cpp:1]: Why do we need `-fno-sycl-early-optimizations`? I'd think we mostly care about what the code is after default optimizations pipeline...

File: sycl/test/check_device_code/ap_fixed.cpp
Comment (aelovikov-intel) [inline sycl/test/check_device_code/ap_fixed.cpp:14]:
```suggestion
SYCL_EXTERNAL auto test_sqrt(sycl::detail::ap_int<13> a) {
  return __spirv_FixedSqrtINTEL<13, 5>(a, false, 2, 2);
```

And I think that would resolve my question above as well.

File: sycl/test/check_device_code/device_asan_no_ctor.cpp
Comment (aelovikov-intel) [inline sycl/test/check_device_code/device_asan_no_ctor.cpp:16]: I think we need to drop changes in this test, at least for this round of changes.

File: sycl/test/check_device_code/device_global_const_eval_use.cpp
Comment (aelovikov-intel) [inline sycl/test/check_device_code/device_global_const_eval_use.cpp:83]: Do we even need this function? If yes, can we put the attribute on the `device_global` variables instead and remove it still?

File: sycl/test/check_device_code/device_has.cpp
Comment (aelovikov-intel) [inline sycl/test/check_device_code/device_has.cpp:58]: I think we might need *both* new and old code, actually. Someone from @intel/dpcpp-tools-reviewers should review changes in this file.

File: sycl/test/check_device_code/fpga_ihs_float.cpp
Comment (aelovikov-intel) [inline sycl/test/check_device_code/fpga_ihs_float.cpp:2]: Drop this change.

File: sycl/test/check_device_code/fpga_ihs_float.cpp
Comment (aelovikov-intel) [inline sycl/test/check_device_code/fpga_ihs_float.cpp:372]: Why not eliminate all the templates similarly to the changes in one of the files above?

File: sycl/test/check_device_code/group_ballot.cpp
Comment (aelovikov-intel) [inline sycl/test/check_device_code/group_ballot.cpp:6]: This test should also be moved out of this folder.

File: sycl/test/check_device_code/matrix/matrix_load_store_as.cpp
Comment (aelovikov-intel) [inline sycl/test/check_device_code/matrix/matrix_load_store_as.cpp:12]: Matrix tests belong to a different team (or at least should) and thus should go to a separate PR.

File: sycl/test/check_device_code/device_global_const_eval_use.cpp
Comment (ianayl) [reply sycl/test/check_device_code/device_global_const_eval_use.cpp:83]: You'll have to excuse me: Which attribute? SYCL_EXTERNAL? As far as I can tell, SYCL_EXTERNAL cannot be applied to variables in the anonymous/global namespace.

File: sycl/test/check_device_code/group_ballot.cpp
Comment (ianayl) [reply sycl/test/check_device_code/group_ballot.cpp:6]: Should the group tests have their own group folder? Or does it just not belong in check_device_code?

File: sycl/test/check_device_code/matrix/matrix_load_store_as.cpp
Comment (ianayl) [reply sycl/test/check_device_code/matrix/matrix_load_store_as.cpp]: According to the CODEOWNERS file, this is _technically_ ours, although that is probably a typo. Either way, I've reverted the changes, and will be making a new PR for the matrix team.

Should I do anything about CODEOWNERS? I will communicate with the matrix team regardless.

File: sycl/test/check_device_code/device_asan_no_ctor.cpp
Comment (ianayl) [reply sycl/test/check_device_code/device_asan_no_ctor.cpp]: I've reverted the changes, although for my understanding's sake, can I ask why?

Comment (aelovikov-intel) [reply sycl/test/check_device_code/device_asan_no_ctor.cpp]: I'm not sure if the transformation works on device functions or device kernels. If the latter, then the modified test didn't really go through the intended code path.

File: sycl/test/check_device_code/device_global_const_eval_use.cpp
Comment (aelovikov-intel) [reply sycl/test/check_device_code/device_global_const_eval_use.cpp:83]: >  As far as I can tell, SYCL_EXTERNAL cannot be applied to variables in the anonymous/global namespace.

That answers my question, and the change here is good.

File: sycl/test/check_device_code/group_ballot.cpp
Comment (aelovikov-intel) [reply sycl/test/check_device_code/group_ballot.cpp:6]: The latter - LLVM IR isn't verified in any way in this test.

File: sycl/test/check_device_code/matrix/matrix_load_store_as.cpp
Comment (aelovikov-intel) [reply sycl/test/check_device_code/matrix/matrix_load_store_as.cpp]: > Should I do anything about CODEOWNERS? I will communicate with the matrix team regardless.

Yes, please create a PR to update it.

File: sycl/test/check_device_code/group_ballot.cpp
Comment (aelovikov-intel) [reply sycl/test/check_device_code/group_ballot.cpp:6]: I'd start with `sycl/test/extensions/group_ballot.cpp`

File: sycl/test/check_device_code/ap_fixed.cpp
Comment (ianayl) [reply sycl/test/check_device_code/ap_fixed.cpp]: Fixed, thanks for the pointer!

File: sycl/test/check_device_code/fpga_ihs_float.cpp
Comment (ianayl) [reply sycl/test/check_device_code/fpga_ihs_float.cpp]: Fixed!

File: sycl/test/check_device_code/fpga_ihs_float.cpp
Comment (ianayl) [reply sycl/test/check_device_code/fpga_ihs_float.cpp]: My bad, this was one of the earlier test cases I did, I didn't know I could do it as templates. It's fixed now!

File: sycl/test/check_device_code/ap_fixed.cpp
Comment (AlexeySachkov) [inline sycl/test/check_device_code/ap_fixed.cpp:11]: Why do we need the whole `sycl/sycl.hpp` here? Especially if we don't use kernel submission API anymore

File: sycl/test/check_device_code/kernel_arguments_as.cpp
Comment (AlexeySachkov) [inline sycl/test/check_device_code/kernel_arguments_as.cpp:38]: This test is specifically about kernel arguments, I'm not sure if rewriting it is correct.

File: sycl/test/check_device_code/ap_fixed.cpp
Comment (ianayl) [reply sycl/test/check_device_code/ap_fixed.cpp]: My bad! My editor was complaining about missing certain SYCL features, so I had assumed that I needed sycl.hpp.

This is fixed now!

File: sycl/test/check_device_code/device_has.cpp
Comment (ianayl) [reply sycl/test/check_device_code/device_has.cpp]: I split the test cases into 2 files, one for kernels and one for functions. Am I good to proceed with this?

File: sycl/test/check_device_code/kernel_arguments_as.cpp
Comment (ianayl) [reply sycl/test/check_device_code/kernel_arguments_as.cpp]: That makes sense, I've gone ahead and reverted the changes.

**Comment by ianayl** [comment]
@AlexeySachkov Please excuse the ping, I've made some changes following your comments: Are these changes good, or do they need further revision?

Thanks in advance!

**Review by AlexeySachkov — APPROVED** [review]

**Comment by ianayl** [comment]
@aelovikov-intel Are these changes good? I believe I still need your approval before I can move onto CI/github workflow. Thanks!

**Comment by aelovikov-intel** [comment]
> @aelovikov-intel Are these changes good? I believe I still need your approval before I can move onto CI/github workflow. Thanks!

These are different approvals. I'll continue my PR review later today.

**Review by aelovikov-intel — APPROVED** [review]

**Comment by ianayl** [comment]
Hey all,

Terribly sorry for the mess here: I was trying to remove certain commits from this branch and didn't know rebasing would cause everyone in a prior merge to be notified. I will be opening up a new branch for this instead.

**Comment by bader** [comment]
@ianayl, please, reset your branch to 88cba5e7b34d5a66bf88a70bc6913a5696794d2d and re-open the pull request.
Avoid using rebase and just put fixes as separate commits on top of your branch.

**Comment by ianayl** [comment]
@bader My bad, that is what I am currently doing. Thanks for your patience!

**Pull request closed without merging**
